metals/src/main/scala/scala/meta/internal/metals/callHierarchy/CallHierarchyItemBuilder.scala

```diff
@@ -0,0 +1,148 @@
+package scala.meta.internal.metals.callHierarchy
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+
+import scala.meta.internal.metals.BuildTargets
+import scala.meta.internal.metals.Compilers
+import scala.meta.internal.metals.HoverExtParams
+import scala.meta.internal.metals.Icons
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.semanticdb.Scala._
+import scala.meta.internal.semanticdb.SymbolOccurrence
+import scala.meta.internal.semanticdb.TextDocument
+import scala.meta.internal.{semanticdb => s}
+import scala.meta.io.AbsolutePath
+import scala.meta.pc.CancelToken
+
+import org.eclipse.{lsp4j => l}
+
+private[callHierarchy] final class CallHierarchyItemBuilder(
+    workspace: AbsolutePath,
+    icons: Icons,
+    compilers: () => Compilers,
+    buildTargets: BuildTargets,
+) {
+  private def buildItem(
+      name: String,
+      kind: l.SymbolKind,
+      uri: String,
+      range: l.Range,
+      selectionRange: l.Range,
+      detail: String,
+      data: CallHierarchyItemInfo,
+  ): l.CallHierarchyItem = {
+    val item = new l.CallHierarchyItem(name, kind, uri, range, selectionRange)
+    item.setDetail(detail)
+    item.setData(data)
+    item
+  }
+
+  private def getClassContructorSymbol(
+      info: s.SymbolInformation
+  ): List[String] =
+    if (info.isClass)
+      info.signature
+        .asInstanceOf[s.ClassSignature]
+        .declarations
+        .flatMap(
+          _.symlinks.find(_.endsWith("`<init>`()."))
+        )
+        .toList ++ Option.when(info.isCase)(
+        info.symbol.replace("#", ".apply().")
+      ) // handling case class constructors
+    else Nil
+
+  /** Get the name of the class of a constructor */
+  private def getName(info: s.SymbolInformation): String =
+    if (info.isConstructor && info.displayName == "<init>")
+      info.symbol.slice(
+        info.symbol.lastIndexOf("/") + 1,
+        info.symbol.lastIndexOf("#"),
+      )
+    else info.displayName
+
+  private def getDetail(
+      signature: String,
+      visited: Array[String],
+      symbol: String,
+  ): String =
+    (if (visited.dropRight(1).contains(symbol)) icons.sync + " "
+     else "") + signature
+
+  private def getSignatureFromHover(hover: Option[l.Hover]): Option[String] =
+    (for {
+      hover <- hover
+      hoverContent <- hover.getContents().asScala.toOption
+      `match` <- """Symbol signature\*\*:\n```scala\n(.*)\n```""".r
+        .findFirstMatchIn(hoverContent.getValue)
+    } yield `match`.group(1))
```

Comment:
Parenthesis around for aren't needed. I wonder, call to `group` might return null, but we are certain that `match` will have that group defined. It can be wrapped into an option but it seems like a redundant "should never happen" case 🤔 

Anyway, I think `""Symbol signature\*\*:\n```scala\n(.*)\n```""".r` can be extracted and enhanced with a good comment.